libraries/AP_Frsky_Telem/frsky.cpp

```diff
@@ -0,0 +1,219 @@
+#include <AP_HAL/AP_HAL.h>
+
+#include "frsky.h"
+
+/*
+   State machine to process incoming Frsky SPort bytes
+*/
+bool FRSky::parse_sport_telemetry_data(uint8_t data, sport_parse_state_t &parse_state)
+{
+    switch (parse_state.state) {
+        case ParseState::START:
+            if (data == FRAME_HEAD) {
+                parse_state.state = ParseState::IN_FRAME ;
+                parse_state.telemetry_rx_buffer_count = 0;
+            } else {
+                if (parse_state.telemetry_rx_buffer_count < TELEMETRY_RX_BUFFER_SIZE) {
+                    parse_state.telemetry_rx_buffer[parse_state.telemetry_rx_buffer_count++] = data;
+                }
+                parse_state.state = ParseState::IN_FRAME;
+            }
+            break;
+
+        case ParseState::IN_FRAME:
+            if (data == FRAME_DLE) {
+                parse_state.state = ParseState::XOR; // XOR next byte
+            } else if (data == FRAME_HEAD) {
+                parse_state.state = ParseState::IN_FRAME ;
+                parse_state.telemetry_rx_buffer_count = 0;
+                break;
+            } else if (parse_state.telemetry_rx_buffer_count < TELEMETRY_RX_BUFFER_SIZE) {
+                parse_state.telemetry_rx_buffer[parse_state.telemetry_rx_buffer_count++] = data;
+            }
+            break;
+
+        case ParseState::XOR:
+            if (parse_state.telemetry_rx_buffer_count < TELEMETRY_RX_BUFFER_SIZE) {
+                parse_state.telemetry_rx_buffer[parse_state.telemetry_rx_buffer_count++] = data ^ STUFF_MASK;
+            }
+            parse_state.state = ParseState::IN_FRAME;
+            break;
+
+        case ParseState::IDLE:
+            if (data == FRAME_HEAD) {
+                parse_state.telemetry_rx_buffer_count = 0;
+                parse_state.state = ParseState::START;
+            }
+            break;
+
+    } // switch
+
+    if (parse_state.telemetry_rx_buffer_count >= SPORT_PACKET_SIZE) {
+        parse_state.telemetry_rx_buffer_count = 0;
+        parse_state.state = ParseState::IDLE;
+        return true;
+    }
+    return false;
+}
+
+/*
+ * check sport packet against its crc
+ * skip first byte (START_STOP)
+ * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
+ */
+bool FRSky::should_process_sport_packet(const uint8_t *packet)
+{
+    return should_process_sport_packet(packet, nullptr);
```

Comment:
This can be accomplished with C++-style default parameters; the header gets an assignment for the second parameter to nullptr.

Reply:
> I'd suggest we have enough of our home-grown bugs to work through; importing new ones seems strange.
> 
> Is there a reason that two consecutive FRAME_HEAD bytes?

mhmm, this is the OpenTX state machine and I assumed it was well tested.  Well formed packets should never have the transition START->frame_head, I can try to simplify the state machine but rather than a bug I think the OpenTX guys hit some bad behaving tx/rx combo, I just don't wan't to reinvent the wheel like I had to do for fport "undocumented" feature vs bug packet rate control

Reply:
`Is there a reason that two consecutive FRAME_HEAD bytes?`
@peterbarker shouldn't we read this as "no matter what you received so far, as soon as you see a FRAME_HEAD reset the state machine" ?